### PR TITLE
Optimization of Qwen3-TTS on Wormhole N150 and N300

### PR DESCRIPTION
**Pull request by ign-rahultp**
# Optimization of Qwen3-TTS on Wormhole N150 and N300

## Summary

Device-performance optimization of **Qwen3-TTS-12Hz-1.7B on Wormhole N150 (1x1) and N300
(1x2, TP=2)**: the CodePredictor and Talker decode paths retuned for the Wormhole 8x8 grid
and 12 DRAM banks, the whole CodePredictor frame fused into one Metal trace, the AR loop's
host round-trips removed, the ECAPA speaker encoder made traceable, plus a full-model PCC
suite, a voice-clone quality gate, per-op Tracy reporting, and Tier 3 CI on WH N150,
WH N300 and BH P150.

Refer to [models/demos/qwen3_tts/README.md](README.md) for how to run it.

## Performance

1. **CodePredictor frame fused into one trace** — prefill, the residual decodes, device
   sampling (top-k + Gumbel) and the codec embeddings replay as a single Metal trace, one
   dispatch instead of one per step. The obsolete 8192-wide top-k pad was dropped.

2. **CodePredictor layer sharded on N300 (TP=2)** — the Talker's sharded RMSNorm,
   `nlp_create_qkv_heads` and `nlp_concat_heads` ported to the CP behind `mesh_utils.is_n300`,
   and the TP all-reduce turned into one all-gather plus a local add of the halves.

3. **CodePredictor layout and collectives** — layout ops stripped out of decode, the
   all-gather handed its input already sharded, `all_gather_async` with caller-owned
   semaphores (`QWEN3_TTS_CCL_ASYNC`), in-model DRAM-sharded grids for gate/up and down
   instead of `find_grid_k_n`, and o_proj DRAM-sharded with its unpad folded into the
   all-reduce.

4. **Fused GQA SDPA in the CodePredictor** (`QWEN3_TTS_CP_FUSED_SDPA`, default ON) — replaces
   the manual fp32 BMM chain (typecast, `repeat_interleave`, QK, scale, mask, softmax, PV).
   The k-chunk is 32 everywhere: the CP KV cache is one tile deep, so 64 padded against
   nothing.

5. **Talker decode** — fused SDPA, 1D gate/up, residual kept sharded, RoPE and the SDPA mask
   hoisted out of the layer, the QKV `InterleavedToSharded` skipped, norms run on their
   consumer's grid, and a decode-layout head split (`QWEN3_TTS_DECODE_HEAD_SPLIT`) that
   removes the per-head transposes.

6. **Decode-mode RoPE** — `is_decode_mode=True` rotates every head inside one tile instead of
   looping once per head. Gated on `seq == 1`, not on the device, so both SKUs get it.

7. **Talker prefill** — width-sharded RMSNorm feeding 1D matmuls with no layout hop between,
   residual adds writing the norm's shard spec, per-bucket NLP memory configs, a DRAM-sharded
   o_proj at short seq, width-sharded gate/up on swept grids, and prefill RoPE folding heads
   into the batch dim. Every swept config is keyed on the exact `(seq, K, N)`, so the two
   SKUs cannot collide.

8. **bfloat8_b decode weights, default ON** (`QWEN3_TTS_BF8_WEIGHTS=0` reverts) — every decode
   matmul is M = 1 tile and weight-bandwidth bound, so fewer weight bytes was the only lever
   left; the one shape not near the DRAM wall (CP gate/up) barely moved.

9. **Fewer host round-trips in the AR loop** — codec 0 sampled inside the Talker trace and
   folded into `tokens_out` so one D2H returns all 16 codes, chip-0-only readback instead of
   `ConcatMeshToTensor`, Gumbel tiles pre-drawn at setup, the trailing-text row uploaded only
   when it changes, and prefill fed by a device-to-device `ttnn.assign`.

10. **ECAPA speaker encoder (time to first audio)** — k=1 TDNNs and 1x1 convs fused into tuned
    device linears, the reflect tap expressed as one permutation matmul, host round-trips cut
    so the forward is traceable, and the mel spectrogram moved into TTNN and captured inside
    that trace. Conv fidelity raised to HiFi4 after the audio gate showed LoFi changed the
    render.

## Testing

New full-model chained PCC suite against production weights:
`tests/test_qwen3_tts_full_model_pcc.py` (speaker encoder -> Talker prefill -> `codec_head` ->
CP prefill + 14 decodes -> next-frame embed -> Talker decode, teacher-forced against the fp32
torch reference, so each stage carries everything upstream of it), with
`test_qwen3_tts_prefill_pcc.py` and `test_qwen3_tts_decode_pcc.py` asserting the same walk one
half at a time, sharing `tests/qwen3_tts_full_model_pcc_common.py`. Thresholds are the lower of
the N150 and N300 measurements, plus a per-frame floor.

New `tests/test_qwen3_tts_voice_quality.py`: speaker similarity (WavLM-SV) and WER
(whisper-large-v3), scored by models independent of the pipeline under test, gating SIM > 0.80
and WER < 0.10 — PCC alone cannot catch a render that has gone runaway.

`tests/test_qwen3_tts_pcc.py` still clears its per-block expectations (`mlp_decode`
0.999693, `attention_decode` 0.999786, `cp_step` 0.999852, `talker_chain` 0.974535 on
N300, each within the suite's 0.005 tolerance of its golden). Perf
instrumentation: `tests/test_qwen3_tts_perf_device.py` (demo prefill and steady AR step) and
`tests/perf/`, two self-capturing Tracy reports — one Talker prefill layer and one decode step
block by block — each writing a ranked report plus a full per-op list with shapes, dtypes,
memory configs and core counts.

## Measured performance — Wormhole N150 and N300

`tests/test_qwen3_tts_perf_device.py`, which runs the demo once and reads the two timings off
the inference pass (Metal trace + KV cache, one command queue, warm kernel cache). Prompt:
"Good morning. Today is a beautiful day for a walk in the park, with bright sun and a gentle
breeze through the trees." — 61 real / 64 padded prefill tokens.

| SKU | Prefill | Steady AR step |
|---|---|---|
| Wormhole N150 (TP=1) | **15.5 - 16.0 ms** | **39.0 - 39.6 ms/frame** |
| Wormhole N300 (TP=2) | **17.4 - 17.6 ms** | **37.6 - 38.4 ms/frame** |

Real time is 80 ms/frame (12.5 codec frames per second), so both SKUs render at roughly RTF
0.47-0.49 in steady state.

## CI

Registered as **Tier 3** on WH N150, WH N300 and BH P150, following
`models/MIGRATING_TO_TIERED_CI.md`:

| Pipeline | Job | Command | Timeout / SKU |
|---|---|---|---|
| e2e | Qwen3-TTS-1.7B e2e tests | `demo_full_ttnn_tts.py --hf-id "$HF_MODEL" --seed 42` then `test_qwen3_tts_voice_quality.py` | 8 min (wh_n150, bh_p150) |
| e2e | Qwen3-TTS-1.7B e2e tests (N300 TP=2) | same, with `MESH_DEVICE=N300` | 8 min (wh_n300) |
| unit | Qwen3-TTS-1.7B unit tests | `test_qwen3_tts_pcc.py` | 3 min (all three SKUs) |

The N300 e2e leg is a separate entry because `MESH_DEVICE` is per-command, not per-SKU; the
unit leg needs none, since it opens a plain single-chip device.

Link to the CI runs:
1. [(Tier 3) Models Unit · tenstorrent/tt-metal@3b47de2](https://github.com/tenstorrent/tt-metal/actions/runs/34439428240)
2. [(Tier 3) Models E2E · tenstorrent/tt-metal@3b47de2](https://github.com/tenstorrent/tt-metal/actions/runs/34439463928)
3. [(Tier 3) Models Unit · tenstorrent/tt-metal@3b47de2](https://github.com/tenstorrent/tt-metal/actions/runs/34439505549)
4. [(Tier 3) Models E2E · tenstorrent/tt-metal@3b47de2](https://github.com/tenstorrent/tt-metal/actions/runs/34439534096)

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ign/qwen3_tts_opt)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ign/qwen3_tts_opt)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ign/qwen3_tts_opt)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ign/qwen3_tts_opt)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ign/qwen3_tts_opt)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ign/qwen3_tts_opt)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ign/qwen3_tts_opt)

### CODE REVIEW

**Comment by ign-rahultp** [comment]
Hi @mtairum 

The Tier 3 CI tests (Models Unit and E2E) are failing on the WH SKUs — wh_n150 and wh_n300 — because the Qwen3-TTS weights are not pre-staged on the runners. This may need the CI/infra owner to stage Qwen/Qwen3-TTS-12Hz-1.7B-Base.
Sharing the CI runs below that were failed because of this error;
1. [(Tier 3) Models Unit · tenstorrent/tt-metal@3b47de2](https://github.com/tenstorrent/tt-metal/actions/runs/34439428240)
2. [(Tier 3) Models E2E · tenstorrent/tt-metal@3b47de2](https://github.com/tenstorrent/tt-metal/actions/runs/34439463928)
3. [(Tier 3) Models Unit · tenstorrent/tt-metal@3b47de2](https://github.com/tenstorrent/tt-metal/actions/runs/34439505549)
4. [(Tier 3) Models E2E · tenstorrent/tt-metal@3b47de2](https://github.com/tenstorrent/tt-metal/actions/runs/34439534096)

Could you please help for getting the weights in the CI..?

**Comment by yieldthought** [comment]
Codex review:

1. **[P1] Use the last real prefill row for the first CodePredictor frame** — [models/demos/qwen3_tts/tt/server.py:2125](https://github.com/tenstorrent/tt-metal/blob/07f399c66ac29b4bd20ae72124d1bbd33664e07e/models/demos/qwen3_tts/tt/server.py#L2125)

   The prefill hidden tensor still has the bucket length here. For the documented 61-token input padded to 64, this selects hidden row 63, while codec 0 was sampled from row 60. The first residual-code frame therefore receives a padding-token hidden state instead of the state that predicted codec 0. The per-step/server path has the same last-row selection in [models/demos/qwen3_tts/tt/utils.py:453](https://github.com/tenstorrent/tt-metal/blob/07f399c66ac29b4bd20ae72124d1bbd33664e07e/models/demos/qwen3_tts/tt/utils.py#L453). The new full-model PCC walk trims to real_seq_len before selecting its last row, so it misses this discrepancy. Suggestion: preserve and pass the last real hidden row from prefill in both generation paths, and exercise that production handoff in the PCC test.

2. **[P1] Finish the decoder worker before collecting final audio** — [models/demos/qwen3_tts/tt/generator.py:126](https://github.com/tenstorrent/tt-metal/blob/07f399c66ac29b4bd20ae72124d1bbd33664e07e/models/demos/qwen3_tts/tt/generator.py#L126)

   The demo calls get_all_audio() before stop(), and an empty token_queue only means the worker has dequeued its last token. If the worker publishes its final chunk after get_nowait() observes an empty audio queue but before the decoded_up_to check, this method skips decoding the remaining tokens and returns without that chunk. A controlled CPU-only schedule against this class reproduced an empty returned audio result with one completed audio chunk still queued. At the demo's chunk size, a lost full chunk is four seconds of audio. Suggestion: drain all submitted tokens, wait for worker completion, then collect output and finalize the partial chunk under a single ownership protocol.

3. **[P2] Keep repetition penalties active with fused sampling** — [models/demos/qwen3_tts/tt/server.py:2061](https://github.com/tenstorrent/tt-metal/blob/07f399c66ac29b4bd20ae72124d1bbd33664e07e/models/demos/qwen3_tts/tt/server.py#L2061)

   With the default fused CP path, --repetition-penalty 1.2 still enables this in-trace codec-0 sampler. _DeviceSampler only uses logits, top-k and Gumbel noise; it receives neither the penalty nor token history. The decode loop consequently bypasses sample_token(), which is where the penalty is implemented, so the advertised option has no effect after prefill. Suggestion: use the existing host codec-0 sampling path whenever repetition_penalty is not 1.0, until the equivalent history-dependent adjustment is implemented on device.

4. **[P2] Honor the exposed trace and KV-cache mode switches** — [models/demos/qwen3_tts/tt/server.py:1381](https://github.com/tenstorrent/tt-metal/blob/07f399c66ac29b4bd20ae72124d1bbd33664e07e/models/demos/qwen3_tts/tt/server.py#L1381)

   The CLI forwards --no-trace and --no-kv-cache here, but use_trace only controls whether 2CQ is disabled, and use_kv_cache only changes the printed mode. The function then allocates KV caches and captures/replays prefill, Talker and CP traces unconditionally. A caller trying to avoid trace-related failures or compare untraced execution still runs the traced cached implementation. Suggestion: implement the requested execution modes, or remove the switches and reject unsupported false values at the boundary rather than silently accepting them.

Verification: manual source review, Python syntax parsing, and CPU-only control-flow probes. No Tenstorrent hardware tests, numerical PCC runs, audio-quality evaluation, or performance measurements were run.

File: models/demos/qwen3_tts/tt/server.py [hunk 1]
```diff
@@ -0,0 +1,3560 @@
...
+    # replaces was the single most expensive thing in the AR loop that was not a trace
+    # replay — a [1,1,1,3072] fp32 D2H plus a host top-k/softmax/multinomial — and it
+    # also forced a per-frame H2D to hand the token back for the next frame's CP trace.
+    # Writing straight into tok_bufs[0], the buffer the fused CP frame reads, removes
+    # all three for the cost of one in-trace topk. Slot 15 of the noise tile is free:
+    # the CP frame uses slots 0-14.
+    #
+    # fused_sampler and fused_tok_bufs are allocated before ANY trace capture (see the
+    # unsafe-allocation note where they are created), so both are safe to bake in here.
+    codec0_in_talker_trace = fused_sampler is not None and not config.greedy
```
Comment (yieldthought) [inline models/demos/qwen3_tts/tt/server.py]: Codex review: [P2] Preserve the requested repetition penalty. The device sampler has no token-history input, so use the existing host codec-0 sampler for non-default penalties. The fused CP frame can remain enabled.

Suggestion:
```suggestion
    codec0_in_talker_trace = fused_sampler is not None and not config.greedy and config.repetition_penalty == 1.0
```

**Comment by ign-rahultp** [comment]
Addressed the Codex review comments @yieldthought

**Comment by mtairum** [comment]
Re-ran the Tier 3 pipelines on the current head (`98909bf1`). The earlier runs failed before reaching the model: the Qwen3-TTS weights were not in the CI HuggingFace cache, so `snapshot_download` raised `LocalEntryNotFoundError` under `HF_HUB_OFFLINE=1`. The weights have now been fetched onto `/mnt/MLPerf/huggingface/hub` manually, and these runs were dispatched with `mlperf-read-only=false` so the TT cache under `/mnt/MLPerf/huggingface/tt_cache/...` can be written on the first pass.

Both runs were dispatched with `model=qwen3-tts-1.7b` and `sku=all`, so the matrix covers every SKU the model registers:

| Pipeline | Run | SKUs in matrix |
|---|---|---|
| (Tier 3) Models Unit | [34606913550](https://github.com/tenstorrent/tt-metal/actions/runs/34606913550) | wh_n150, wh_n300, bh_p150 |
| (Tier 3) Models E2E | [34606919874](https://github.com/tenstorrent/tt-metal/actions/runs/34606919874) | wh_n150, bh_p150, and wh_n300 (TP=2) |

Note that bh_p150 was not exercised by the previous attempt, so this is the first run for that SKU.

File: models/demos/audio/whisper/demo/run_jp_stt.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    if sr != 16000:
+        from scipy.signal import resample
+
+        n = int(round(audio.shape[0] * 16000 / float(sr)))
+        audio = resample(audio, n).astype(np.float32)
+        sr = 16000
+    return sr, audio
+
+
+def main() -> int:
```
Comment (atupe-tt) [inline models/demos/audio/whisper/demo/run_jp_stt.py:50]: @ign-rahultp Can we add these tests to the main `demo.py` by parameterizing the pytest? Since the only difference is the input audio

Comment (ign-rahultp) [reply models/demos/audio/whisper/demo/run_jp_stt.py:50]: @atupe-tt `run_jp_stt.py` was added by TT Admin in 3f1cf82c933 ([demos] Japanese Whisper STT + Qwen3-TTS on Wormhole N150); the only change on this branch is the SPDX header Cycode required. Should we fold it into demo.py in this PR, or leave the Whisper demo as it is..?

**Comment by ign-rahultp** [comment]
CI passed with the latest changes. Refer to

1. [(Tier 3) Models Unit](https://github.com/tenstorrent/tt-metal/actions/runs/34979509590)
2. [(Tier 3) Models E2E](https://github.com/tenstorrent/tt-metal/actions/runs/34979553248)

**Review by mrakitaTT — CHANGES_REQUESTED** [review]
Hi @ign-rahultp, thank you for contributing to our codebase! We really appreciate community contributions, and I don't want to discourage you from contributing.

That said, we already have a team working on the same set of audio models, so we need to coordinate our efforts to avoid duplicating implementations. For example, we had already implemented, optimized, and tested XTTS-v2 and integrated it into tt-inference-server. Another implementation of the same model was later merged by you into tt-metal under a different directory name (xtts instead of xtts_v2). We'd like to avoid ending up in a similar situation going forward.

We're currently working on [Qwen3-TTS](https://github.com/tenstorrent/tt-metal/pull/57159) as well. Our implementation is more compact with half less code, supports all model variants, and currently has better performance. Given the overlap, I'll need to block this PR so that we don't end up maintaining two implementations of the same model in different locations.

If there's functionality in your implementation that our current implementation is missing, please feel free to submit additional PR that adds it on top of our implementation. We'd be happy to review it.

For future work, it would also be helpful to understand what you're looking to achieve. Are you primarily interested in simply making contributions to the project, or are there specific models that you or your company need to run on our hardware?

Either way, feel free to share the list of models you're planning to work on. We can coordinate with you and let you know which ones are not already being worked on by our team. That should help us avoid duplicated effort and make your contributions easier to integrate.
